### ISSUE TITLE
Output pulse timing error when using CBox_v3

### ISSUE THREAD

**Issue by gtaifu** [DiCarloLab-Delft/PycQED_py3#88]
When using CBox_v3 to perform experiments, Adriaan encountered the problem that the pulse timing is not exactly the same as what is defined in the QuMIS instructions. ( @AdriaanRol If you think it is necessary, you could put your code and figure here)

After having a discussion with Jacob and Adriaan, Xiang decided to solve this problem in three phases.

## Phase 1
**Time**: Before Jan. 9th. 
Xiang will look into the experiment data got by Adriaan and then think about the possible reasons and try to identify which one would be the cause. Also, Xiang will think about possible experiments to exclude the possible causes. The same FPGA firmware will be used since the error would be built-dependent.  Build-dependent means that if you the same thing twice, sometimes you will see it, but sometimes not.

## Phase 2
**Time**: The week of Jan. 9th and early of the following week.
**Goal**: to reproduce the bugs and trying to figure out where it is.
 
Xiang will try to reproduce the problem without connecting to the fridge at first. Adriaan will help Xiang with offering the test case.

If this does not reproduce the bug, then the problem may relate to the setup. In this week, Xiang will stay in the lab most, when Adriaan reproduce the bug, Xiang will work on that.

Every day in the morning, Xiang, Adriaan and Jacob will have a **short** meeting discussing the current situation and following plan. Most likely, Jacob will stay in the high-level offering suggestions.
 
## Phase 3
**Time**: The week of Jan. 16th.
**Goal**: To solve the problem.
Does the problem reside in the CBox_v3?
- Yes. The bug can be solved by Xiang without connecting to the setup independently.
- No. 
  - CBox_v3 is interfered by other devices;
  - The interconnection in the setup is problematic;
  - Specific command order would destroy the internal state of the CBox;
  - ...

As mentioned by Jacob, the problem may arise from a ridiculous source. We need to be careful with that.

## NOTE
A Mind map-like tool will be used to sort the potential problem.

**Comment by gtaifu** [comment]
By firstly uploading a .qumis file used for RB which is of around 8000 lines, then execute the following commands:
``` Python
CBox.demodulation_mode('double')
sleeptime = 2
for i in range(3):
    CBox.acquisition_mode('idle')
    CBox.core_state('idle')
    CBox.run_mode('idle')
    CBox.load_instructions('programs\\NoProblem.qumis')
    CBox.acquisition_mode('integration averaging')
    CBox.core_state('active')
    CBox.run_mode('run')
    print('NoProblem code: %d' % i)
    time.sleep(sleeptime)
    CBox.acquisition_mode('idle')
    CBox.core_state('idle')
    CBox.run_mode('idle')
    CBox.load_instructions('programs\\ErrorReproduce.qumis')
    CBox.acquisition_mode('integration averaging')
    CBox.core_state('active')
    CBox.run_mode('run')
    print('ErrorReproduce code: %d' % i)
    time.sleep(sleeptime)
``` 
NoProblem.qumis:
``` asm
mov r0, 20000   # r0 stores the cycle time , 100 us
mov r1, 0       # sets the inter pulse wait to 0
mov r14, 0      # r14 stores number of repetitions
# Experiment: repeat the rounds for infinite times
Exp_Start:

WaitReg r0
WaitReg r0
wait 8
pulse 0000 1111 1111
wait 40
measure

WaitReg r0
WaitReg r0
trigger 0000000, 2
wait 2
trigger 1000001, 2
wait 2
wait 8
pulse 0000 1111 1111
wait 40
measure
beq r14, r14, Exp_Start       # Infinite loop
```
ErrorReproduce.qumis:
``` asm
mov r0, 20000   # r0 stores the cycle time , 100 us
mov r1, 0       # sets the inter pulse wait to 0
mov r14, 0      # r14 stores number of repetitions
# Experiment: repeat the rounds for infinite times
Exp_Start:

WaitReg r0
WaitReg r0
wait 8
pulse 0000 1111 1111
wait 40
measure
wait 400

WaitReg r0
WaitReg r0
trigger 0000000, 2
wait 2
trigger 1000001, 2
wait 2
wait 8
pulse 0000 1111 1111
wait 40
measure
wait 400
beq r14, r14, Exp_Start       # Infinite loop
```

I got the following output:
![image](https://cloud.githubusercontent.com/assets/6672150/22076253/a115aa7e-ddaf-11e6-88e2-774b8fa458aa.png)

It consists of the following components:
![image](https://cloud.githubusercontent.com/assets/6672150/22076268/adab61b6-ddaf-11e6-9604-d73f9e614056.png)
![image](https://cloud.githubusercontent.com/assets/6672150/22076274/b13458e2-ddaf-11e6-9fa6-10febba0f63f.png)
![image](https://cloud.githubusercontent.com/assets/6672150/22076277/b395ac62-ddaf-11e6-928c-db54725d30b3.png)
![image](https://cloud.githubusercontent.com/assets/6672150/22076278/b632a75e-ddaf-11e6-9867-17bd0e04306d.png)

I can make one conclusion and one guess from this observation:
1. CBox_v3 v3.3.1 indeed has bugs. It is not stable enough. I will think the problem confronting Adriaan comes from CBox_v3, instead of the setup.
2. The error **may** come from that the processor runs into the instructions that do not belong to the program. Not sure

**Comment by gtaifu** [comment]
To make sure the reason is point 2 or not, I changed all pulse instructions in the RB.qumis into **pulse 1001 1001 1001**. Then I rerun the NoProblem.qumis:
![image](https://cloud.githubusercontent.com/assets/6672150/22078076/20487ffe-ddb7-11e6-9d78-282f799304d9.png)

ErrorReproduce.qumis:
![image](https://cloud.githubusercontent.com/assets/6672150/22078085/27d7fe5c-ddb7-11e6-91da-e32007b5cff5.png)

The wrong pulse is not pulse **1**!

So I can make the conclusion: the bug comes from that the processor runs into the instructions that do not belong to the program.

**Comment by gtaifu** [comment]
The bug comes from that one instruction after the branch instruction is not correctly cleared. It is Solved in [CBox v3.3.2](https://github.com/DiCarloLab-Delft/250MSps_ControlBox/releases/tag/code_V3.3.2).

This issue could be closed after the setup test passes.

**Comment by gtaifu** [comment]
In the following test, we found this bug is still there and the cause is the buggy control logic in the pipeline register IF/ID. A new FSM is designed for it in [CBox_v3.3.5](https://github.com/DiCarloLab-Delft/250MSps_ControlBox/releases/tag/code_v3.3.5). Hardware test passed.
